src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+     * @param numberOfBytes the number of bytes after `index` that need to be present.
+     * @return false if not enough bytes were available in the stream to satisfy the request; otherwise, true.
+     */
+    private boolean fillAt(long index, long numberOfBytes) {
+        long shortfall = numberOfBytes - availableAt(index);
+        if (shortfall > 0) {
+            refillableState.bytesRequested = numberOfBytes + (index - offset);
+            if (ensureCapacity(refillableState.bytesRequested)) {
+                // Fill all the free space, not just the shortfall; this reduces I/O.
+                refill(freeSpaceAt(limit));
+                shortfall = refillableState.bytesRequested - availableAt(offset);
+            } else {
+                // The request cannot be satisfied, but not because data was unavailable. Return normally; it is the
+                // caller's responsibility to recover.
+                shortfall = 0;
+            }
+        }
+        if (shortfall <= 0) {
+            refillableState.bytesRequested = 0;
+            refillableState.state = State.READY;
+            return true;
+        }
+        refillableState.state = State.FILL;
+        return false;
+    }
+
+    /**
+     * Moves all buffered (but not yet read) bytes from 'buffer' to the destination buffer.
+     * @param destinationBuffer the destination buffer, which may be 'buffer' itself or a new buffer.
+     */
+    private void moveBytesToStartOfBuffer(byte[] destinationBuffer, int fromIndex) {
+        long size = availableAt(fromIndex);
+        if (size > 0) {
+            System.arraycopy(buffer, fromIndex, destinationBuffer, 0, (int) size);
+        }
+        if (fromIndex > 0) {
+            shiftIndicesLeft(fromIndex);
+        }
+        offset = 0;
+        limit = size;
+    }
+
+    /**
+     * @return the number of bytes that can be written at the end of the buffer.
+     */
+    private long freeSpaceAt(long index) {
+        return refillableState.capacity - index;
+    }
+
+    /**
+     * Reads a single byte without adding it to the buffer. Used when skipping an oversized value, in cases where
+     * the byte values are important (e.g. within the header of the oversized value, in order to determine
+     * the number of bytes to skip).
+     * @return the next byte, or -1 if the stream is at its end.
+     */
+    private int readByteWithoutBuffering() {
+        int b = -1;
+        try {
+            b = refillableState.inputStream.read();
+        } catch (IOException e) {
+            throwAsIonException(e);
+        }
+        if (b >= 0) {
+            refillableState.individualBytesSkippedWithoutBuffering += 1;
+        }
+        return b;
+    }
+
+    /**
+     * Peek at the next byte from the stream, assuming it will be buffered unless the current value is being skipped.
+     * @return the byte, or -1 if the end of the stream has been reached.
+     */
+    private int slowPeekByte() {
```

Comment:
There are a lot of mentions of "slow" mode. I think I understand what that means, but adding an explanation comment early in the file would probably be helpful.

Reply:
Here's the comment on line 477:

```
    /*
     * This class contains methods with the prefix 'slow'. These methods often
     * have prefix-free 'quick' variants that are faster because they are always
     * called from contexts where the buffer is already known to hold all bytes
     * that will be required or available to complete the call, and therefore no
     * filling is required. Sometimes a 'slow' method does not have a 'quick'
     * variant. This is typically because the 'quick' variant is simple enough
     * to be written inline at the call site. Public API methods will delegate
     * to either the 'slow' or 'quick' variant depending on whether 'isSlowMode'
     * is true. In general, 'isSlowMode' may be disabled whenever the input source
     * is a byte array (and therefore cannot grow and does not need to be filled),
     * or whenever the current value's parent container has already been filled.
     * Where a choice exists, 'slow' methods must call other 'slow' methods, and
     * 'quick' methods must call other 'quick' methods.
     */
```

Let me know if you think that's insufficient.